desktop/cmp/form/inputs/Select.js

```diff
@@ -0,0 +1,321 @@
+
+        /** Function to return message indicating no options loaded. Passed current query input. */
+        noOptionsMessageFn: PT.func,
+
+        /**
+         * Preset list of options for selection. Objects must contain a `value` property; a `label`
+         * property will be used for the default display of each option. Other types will be taken
+         * as their value directly and displayed via toString().  See also `queryFn` to  supply
+         * options via an async query (i.e. from the server) instead of up-front in this prop.
+         */
+        options: PT.array,
+
+        /** Text to display when control is empty. */
+        placeholder: PT.string,
+
+        /**
+         * Async function to return a list of options for a given query string input.
+         * Replaces the `options` prop - use one or the other.
+         */
+        queryFn: PT.func,
+
+        /**
+         * Escape-hatch props passed directly to react-select. Use with care - not all props
+         * in the react-select API are guaranteed to be supported by this Hoist component,
+         * and providing them directly can interfere with the implementation of this class.
+         */
+        rsOptions: PT.object
+    };
+
+    baseClassName = 'xh-select';
+
+    // Normalized collection of selectable options. Passed directly to synchronous select.
+    // Maintained for (but not passed to) async select to resolve value string <> option objects.
+    @observable.ref internalOptions = [];
+    @action setInternalOptions(options) {this.internalOptions = options}
+
+    // Prop flags that switch core behavior.
+    get asyncMode() {return !!this.props.queryFn}
+    get creatableMode() {return !!this.props.enableCreate}
+    get multiMode() {return !!this.props.enableMulti}
+
+    constructor(props) {
+        super(props);
+        this.addReaction({
+            track: () => this.props.options,
+            run: (opts) => {
+                opts = this.normalizeOptions(opts);
+                this.setInternalOptions(opts);
+            },
+            fireImmediately: true
+        });
+    }
+
+    render() {
+        const {props, renderValue} = this;
+
+        let reactSelectProps = {
+            value: renderValue,
+
+            autoFocus: props.autoFocus,
+            isDisabled: props.disabled,
+            isMulti: props.enableMulti,
+            menuPlacement: withDefault(props.menuPlacement, 'auto'),
+            menuPortalTarget: document.body,
+            noOptionsMessage: this.noOptionsMessageFn,
+            placeholder: withDefault(props.placeholder, 'Select...'),
+            tabIndex: props.tabIndex,
+
+            classNamePrefix: 'xh-select',
+            styles: this.getStylesConfig(),
+            theme: this.getThemeConfig(),
+
+            onChange: this.onSelectChange
+        };
+
+        if (this.asyncMode) {
+            reactSelectProps = {
+                ...reactSelectProps,
```

Comment:
Why use the spread in these instances, creating 3 more clones.  We already create a new object on line 122, just write into it?